src/pages/blog/2024-08-14-semantic-nullability-for-application-developers.mdx

```diff
@@ -0,0 +1,93 @@
+---
+title: "Semantic Nullability For Application Developers"
+tags: ["spec"]
+date: 2024-09-12
+byline: Alex Reilly
+---
+
+# Semantic Nullability
+
+> This blog post is directed at application developers using GraphQL. If you are a library author, you should read the more detailed feature spec instead.
+
+Today we're providing a progress update from the Nullablility Working Group on Semantic Nullability which is our new approach to fixing GraphQL's nullability issues.
+
+GraphQL has some fundamental problems, and to talk about them, we first have to talk about GraphQL's type system. GraphQL allows you to define a schema, and to do that many developers write a document in Schema Definition Language, or SDL. A SDL document may look like this
+
+```graphql
+type User { 
+  id: ID! 
+  name: String!
+  age: Int
+  posts: [Post] 
+}
+```
+
+One thing to make clear: this is not like using a type system in a compiled language (TS, Swift, Kotlin) where the type system makes compile-time guarantees about behavior at runtime. Rather you can think of GraphQL's type system as a "runtime" type system. You can trust that `age` will be an `Int`, because at runtime, GraphQL will assert that it is a `Int`, and throw an error if it is not. All types effectively represent a typecast, or a type assertion.
+
+This results in `null` having two different meanings in GraphQL.
+1. No value was provided. The `User` never provided an `age`.
+2. There was an error resolving the field. The `User` provided an age in the form of a `Float` and it couldn't be cast to an `Int`, or our new `BirthdayBoy` provider [microservice](https://www.youtube.com/watch?v=y8OnoxKotPQ) timed out and never returned the `User`'s birthday.
+
+We can differentiate between the two by calling the first `null`, and the second `(Error, null)`. The first type is signaled by a `null` return value and no associated error in the errors array. The second is signaled by a `null` return value and an associated error in the errors array.
+
+Returning to the GraphQL type system, we can see it has two options to indicate the nullability of a field.
+1. `String` which we now know means the field can be a `String`, `null`, or `(Error, null)`
+2. `String!` which we now know means the field can be `String`
+
+When an error occurs resolving a `String` field, it's not much of a problem for clients. They can decide if they can deal with that field missing or not. However when an error occurs resolving a `String!` field, GraphQL responds by destroying part of the result data before it's sent to the client. Given the danger, many developers choose to never use any non-nullable fields.
```

Comment:
I think you're meaning the server-side "don't make non-null fields available" idea, rather than the client-side "don't consume non-null fields".

```suggestion
When an error occurs resolving a `String` field, it's not much of a problem for clients. They can decide if they can deal with that field missing or not. However when an error occurs resolving a `String!` field, GraphQL responds by destroying part of the result data before it's sent to the client. Given the danger, many developers choose to never add non-nullable fields to the schema.
```